### PR TITLE
feat: ratelimit - ratelimit refactor + tests

### PR DESCRIPTION
**Pull request by sean-brydon**
This PR adds ratelimiting to booking page. Allowing 10/60s (we can reduce this if required but its better than no support)

This PR also adds a new helper function that throws and provides a detailed error message to the user about what went wrong if they hit the limit. 

How to test
* Add upstash env DEV keys from bitwarden
* try hit any of the rate limited APIs (try booking more than 10 times in 60 seconds)
* Expect error

### CODE REVIEW

File: packages/lib/checkRateLimitAndThrowError.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  if (remaining < 0) {
+    const convertToSeconds = (ms: number) => Math.floor(ms / 1000);
+    const secondsToWait = convertToSeconds(reset - Date.now());
+    throw new TRPCError({
+      code: "TOO_MANY_REQUESTS",
+      message: `Rate limit exceeded. Try again in ${secondsToWait} seconds.`,
+    });
+  }
+}
```
Comment (sean-brydon) [inline packages/lib/checkRateLimitAndThrowError.ts:20]: Helper function to have standardized error messages with ratelimiting

File: packages/lib/rateLimit.ts [hunk 2]
```diff
@@ -6,17 +6,38 @@ import logger from "./logger";
...
+   * How many requests the user has left within the current window.
+   */
+  remaining: number;
+  /**
+   * Unix timestamp in milliseconds when the limits are reset.
+   */
+  reset: number;
+
+  pending: Promise<unknown>;
+};
```
Comment (sean-brydon) [inline packages/lib/rateLimit.ts:33]: This should really be exported by upstash but isnt :(

**Comment by baileypumfleet** [comment]
Can we fix the types, linting and such so tests pass?

File: apps/web/pages/api/auth/forgot-password.ts [hunk 3]
```diff
@@ -34,15 +33,11 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
 
   // 10 requests per minute
 
-  const limiter = await rateLimiter();
-  const limit = await limiter({
+  await checkRateLimitAndThrowError({
```
Comment (zomars) [inline apps/web/pages/api/auth/forgot-password.ts:36]: Noice Thats Nice GIF

File: packages/lib/checkRateLimitAndThrowError.test.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import { describe, expect, it, vi } from "vitest";
+
+import { checkRateLimitAndThrowError } from "./checkRateLimitAndThrowError";
+import { rateLimiter } from "./rateLimit";
+import type { RatelimitResponse } from "./rateLimit";
+
+vi.mock("./rateLimit", () => {
```
Comment (zomars) [inline packages/lib/checkRateLimitAndThrowError.test.ts:7]: Noice Thats Nice GIF

File: packages/lib/checkRateLimitAndThrowError.ts (hunk 1 above)
Comment (zomars) [reply packages/lib/checkRateLimitAndThrowError.ts:20]: Noice Thats Nice GIF

**Review by zomars — APPROVED** [review]
Awesome work @sean-brydon loving the tests as well and overall nice abstraction. 🙏🏽

**Pull request merged**